### PR TITLE
Add and/or compute functions

### PR DESCRIPTION
**Pull request by AdamGS**

### CODE REVIEW

File: vortex-array/src/array/bool/mod.rs [hunk 1]
```diff
@@ -71,6 +71,21 @@ impl BoolArray {
         let buffer = BooleanBuffer::from(bools);
         Self::try_new(buffer, validity).unwrap()
     }
+
+    pub fn true_count(&self) -> usize {
```
Comment (robert3005) [inline vortex-array/src/array/bool/mod.rs]: you can get true_count from the stats?

Comment (AdamGS) [reply vortex-array/src/array/bool/mod.rs]: true

Comment (gatesn) [reply vortex-array/src/array/bool/mod.rs]: count

File: vortex-array/src/array/bool/mod.rs [hunk 2]
```diff
@@ -71,6 +72,10 @@ impl BoolArray {
...
+        self.statistics().compute_true_count().unwrap()
```
Comment (gatesn) [inline vortex-array/src/array/bool/mod.rs]: You can't unwrap

File: vortex-datafusion/src/eval.rs [hunk 3]
```diff
@@ -7,35 +7,84 @@ use vortex::{
...
-                let lhs = ExpressionEvaluator::eval(array.clone(), expr.left.as_ref())?;
-                let rhs = ExpressionEvaluator::eval(array, expr.right.as_ref())?;
-
                 // TODO(adamg): turn and/or into more general compute functions
                 match expr.op {
                     DFOperator::And => {
+                        let lhs = ExpressionEvaluator::eval(array.clone(), expr.left.as_ref())?;
                         let lhs = lhs.into_bool()?;
+
+                        if lhs.true_count() == 0 {
```
Comment (gatesn) [inline vortex-datafusion/src/eval.rs]: You should create and/or compute functions to abstract away these short-circuits.

File: vortex-datafusion/src/eval.rs [hunk 4]
```diff
@@ -7,35 +7,84 @@ use vortex::{
...
-                    DFOperator::GtEq => compare(&lhs, &rhs, Operator::Gte),
-                    DFOperator::Lt => compare(&lhs, &rhs, Operator::Lt),
-                    DFOperator::LtEq => compare(&lhs, &rhs, Operator::Lte),
-                    DFOperator::NotEq => compare(&lhs, &rhs, Operator::NotEq),
+                    DFOperator::Eq => {
+                        let lhs = ExpressionEvaluator::eval(array.clone(), expr.left.as_ref())?;
+                        let rhs = ExpressionEvaluator::eval(array, expr.right.as_ref())?;
+                        compare(&lhs, &rhs, Operator::Eq)
+                    }
+                    DFOperator::Gt => {
```
Comment (gatesn) [inline vortex-datafusion/src/eval.rs]: You're gonna want that `From<DFOperator>` impl!

**Review by gatesn — CHANGES_REQUESTED** [review]

File: vortex-array/src/array/bool/mod.rs (hunk 2 above)
Comment (AdamGS) [reply vortex-array/src/array/bool/mod.rs]: Dropped it, I think its a premature optimization here

File: vortex-datafusion/src/eval.rs (hunk 3 above)
Comment (AdamGS) [reply vortex-datafusion/src/eval.rs]: created new ones that are also more correct IMO

File: vortex-datafusion/src/eval.rs (hunk 4 above)
Comment (AdamGS) [reply vortex-datafusion/src/eval.rs]: I think we should rethink the whole `Operator` thing here, if its more of an `Operation` we can add boolean operations and get rid of the negation logic.

File: vortex-array/Cargo.toml [hunk 5]
```diff
@@ -25,6 +25,8 @@ arrow-cast = { workspace = true }
 arrow-select = { workspace = true }
 arrow-schema = { workspace = true }
 arrow-ord = { workspace = true }
+arrow-arith = { workspace = true }
+bytes = "1"
```
Comment (gatesn) [inline vortex-array/Cargo.toml]: Use workspace version?

File: vortex-array/src/compute/boolean.rs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+pub trait AndFn {
+    fn and(&self, array: &Array) -> VortexResult<Array>;
+}
+
+pub trait OrFn {
+    fn or(&self, array: &Array) -> VortexResult<Array>;
+}
+
+pub fn and(lhs: &Array, rhs: &Array) -> VortexResult<Array> {
+    if let Some(selection) = lhs.with_dyn(|lhs| lhs.and().map(|lhs| lhs.and(rhs))) {
```
Comment (gatesn) [inline vortex-array/src/compute/boolean.rs:22]: You could also try inverting it, see if the rhs implements `and`?

File: vortex-array/src/compute/boolean.rs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        return selection;
+    }
+
+    let lhs = lhs.clone().into_bool()?.into_canonical()?.into_arrow();
```
Comment (gatesn) [inline vortex-array/src/compute/boolean.rs]: The fallback should be `lhs.clone().into_bool().and(rhs)`

This implementation, using arrow_arith, should live alongside BoolArray

File: vortex-array/src/compute/mod.rs [hunk 8]
```diff
@@ -7,6 +7,7 @@
 //! implementations of these operators, else we will decode, and perform the equivalent operator
 //! from Arrow.
 
+use boolean::{AndFn, OrFn};
```
Comment (gatesn) [inline vortex-array/src/compute/mod.rs]: Can you make this match the others? Where AndFn is exported, not the "boolean" module

**Review by gatesn — DISMISSED** [review]

File: Cargo.toml [hunk 9]
```diff
@@ -116,3 +115,6 @@ warnings = "deny"
 [workspace.lints.clippy]
 all = { level = "deny", priority = -1 }
 or_fun_call = "deny"
+
+[profile.release]
+debug = true
```
Comment (gatesn) [inline Cargo.toml]: FYI the "bench" profile is release + debug=true

File: vortex-array/src/array/constant/compute.rs [hunk 10]
```diff
@@ -59,6 +68,45 @@ impl SearchSortedFn for ConstantArray {
...
+                Some(b) => Scalar::bool(b, Nullability::Nullable),
+                None => Scalar::null(constant_array.dtype().as_nullable()),
+            };
+
+            Ok(ConstantArray::new(scalar, constant_array.len()).into_array())
+        } else {
+            AndFn::and(&constant_array.clone().into_bool()?, other)
+        }
+    } else {
+        vortex_bail!("Boolean operations aren't supported on arrays of different lengths")
```
Comment (gatesn) [inline vortex-array/src/array/constant/compute.rs]: This is the style point we were discussing 😅 

I'd rather invert the if statement and bail at the start of the function. That way the bail is kept close to the logic that decides it, and the rest of the function doesn't need to be indented by another 4 spaces.

File: vortex-array/src/array/constant/compute.rs [hunk 11]
```diff
@@ -59,6 +68,45 @@ impl SearchSortedFn for ConstantArray {
...
+fn constant_array_bool_impl(
+    constant_array: &ConstantArray,
+    other: &Array,
+    bool_op: impl Fn((bool, bool)) -> bool,
+) -> VortexResult<Array> {
+    if constant_array.dtype().is_boolean()
+        && other.dtype().is_boolean()
+        && constant_array.len() == other.len()
+    {
+        if let Ok(array) = ConstantArray::try_from(other.clone()) {
```
Comment (gatesn) [inline vortex-array/src/array/constant/compute.rs]: This should use stats instead. Something like:
```
if let Some(true) = other.stats().get_is_constant() {
    let rhs = scalar_at(other, 0)?;
    ...
} else {
   // We don't know if it's constant or not, fallback
}
```

File: vortex-array/src/array/varbin/builder.rs [hunk 12]
```diff
@@ -47,9 +46,9 @@ impl<O: NativePType> VarBinBuilder<O> {
         self.validity.append_null();
     }
 
-    pub fn finish(&mut self, dtype: DType) -> VarBinArray {
-        let offsets = PrimitiveArray::from(mem::take(&mut self.offsets));
-        let data = PrimitiveArray::from(mem::take(&mut self.data));
+    pub fn finish(mut self, dtype: DType) -> VarBinArray {
+        let offsets = PrimitiveArray::from(self.offsets);
+        let data = PrimitiveArray::from(Vec::from(self.data.freeze()));
```
Comment (gatesn) [inline vortex-array/src/array/varbin/builder.rs]: This isn't zero copy is it? We can add a function to PrimitiveArray to construct directly from a buffer (for PType==u8?)

Comment (gatesn) [reply vortex-array/src/array/varbin/builder.rs]: Alternatively, @robert3005 did we discuss VarBin data being a buffer instead of a child array?

File: vortex-array/src/compute/boolean.rs [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+pub trait AndFn {
+    fn and(&self, array: &Array) -> VortexResult<Array>;
+}
+
+pub trait OrFn {
+    fn or(&self, array: &Array) -> VortexResult<Array>;
+}
+
+pub fn and(lhs: &Array, rhs: &Array) -> VortexResult<Array> {
+    if let Some(selection) = lhs.with_dyn(|lhs| lhs.and().map(|lhs| lhs.and(rhs))) {
```
Comment (gatesn) [inline vortex-array/src/compute/boolean.rs:22]: Maybe leave a couple of comments, these functions are hard to parse at first glance

File: vortex-array/src/array/constant/compute.rs [hunk 14]
```diff
@@ -59,6 +68,45 @@ impl SearchSortedFn for ConstantArray {
...
+            let lhs = constant_array.scalar().value().as_bool()?;
```
Comment (gatesn) [inline vortex-array/src/array/constant/compute.rs]: Maybe first check if either scalar is_null, and then I think the more canonical conversion would be `bool::try_from(array.scalar())?`

Comment (gatesn) [reply vortex-array/src/array/constant/compute.rs]: We should chat through the scalar API next week, it's a bit weird, you might have some ideas to improve

File: vortex-array/src/array/constant/compute.rs [hunk 15]
```diff
@@ -59,6 +68,45 @@ impl SearchSortedFn for ConstantArray {
...
+            let rhs = array.scalar().value().as_bool()?;
+
+            let scalar = match lhs.zip(rhs).map(bool_op) {
+                Some(b) => Scalar::bool(b, Nullability::Nullable),
+                None => Scalar::null(constant_array.dtype().as_nullable()),
+            };
+
+            Ok(ConstantArray::new(scalar, constant_array.len()).into_array())
+        } else {
+            AndFn::and(&constant_array.clone().into_bool()?, other)
```
Comment (gatesn) [inline vortex-array/src/array/constant/compute.rs]: Either this fallback will fail (into_bool when it's not?), or into_bool does an expensive expansion of the constant.

I think the best thing is to have the convention that constant goes on the RHS, so we should fallback to `and(other, constant_array)` to allow the RHS encoding a chance to be efficient, e.g. if it's run-end it only needs to run over the values.

File: vortex-array/src/array/varbin/builder.rs (hunk 12 above)
Comment (robert3005) [reply vortex-array/src/array/varbin/builder.rs]: I think we thought about it but then we wanted to have something like ZstdEncoding

Comment (gatesn) [reply vortex-array/src/array/varbin/builder.rs]: I’m starting to think general purpose compression can be configured on buffers at write-time though; using the layouts mechanism.

File: vortex-array/src/array/constant/compute.rs (hunk 15 above)
Comment (AdamGS) [reply vortex-array/src/array/constant/compute.rs]: ~I can't actually fallback into `AndFn` here because this might also be an `or`, I'm trying to think of a better way of generalizing boolean ops here (I do agree with trying to let `rhs` to give us a more efficient implementation)~ Found a solution I'm happy with here.

File: vortex-array/src/array/varbin/builder.rs (hunk 12 above)
Comment (AdamGS) [reply vortex-array/src/array/varbin/builder.rs]: Not zero-copy, but still pretty cheap IMO. Constructing things from `Bytes` is hard because there's no guarantee the instance is exclusive

File: vortex-array/src/array/constant/compute.rs (hunk 10 above)
Comment (AdamGS) [reply vortex-array/src/array/constant/compute.rs]: Changed the function to fail early, I'm willing to concede that here it looks better :)

File: vortex-array/src/compute/boolean.rs (hunk 13 above)
Comment (AdamGS) [reply vortex-array/src/compute/boolean.rs:22]: added a couple in #488

File: vortex-array/Cargo.toml [hunk 16]
```diff
@@ -55,6 +60,7 @@ build-vortex = { path = "../build-vortex" }
 
 [dev-dependencies]
 criterion = { workspace = true }
+rstest = "0.21"
```
Comment (robert3005) [inline vortex-array/Cargo.toml]: put this in the top level Cargo.toml

File: vortex-array/src/array/bool/compute/boolean.rs [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+use crate::compute::{AndFn, OrFn};
+use crate::{Array, ArrayData, IntoArray, IntoCanonical};
+
+impl OrFn for BoolArray {
+    fn or(&self, array: &Array) -> VortexResult<Array> {
+        let lhs = self.clone().into_canonical()?.into_arrow();
+        let lhs = lhs.as_boolean();
+
+        let rhs = array.clone().into_canonical()?.into_arrow();
+        let rhs = rhs.as_boolean();
```
Comment (robert3005) [inline vortex-array/src/array/bool/compute/boolean.rs:16]: Should this maybe dispatch on the right hand side instead of converting to arrow?

File: vortex-array/src/array/constant/compute.rs [hunk 18]
```diff
@@ -59,11 +70,59 @@ impl SearchSortedFn for ConstantArray {
...
+        vortex_bail!("Boolean operations are only supported on boolean arrays")
+    }
+
+    // If the right side is constant
+    if let Some(true) = other.statistics().compute_is_constant() {
+        let lhs = constant_array.scalar().value().as_bool()?;
+        let rhs = scalar_at(other, 0)?.value().as_bool()?;
+
+        let scalar = match lhs.zip(rhs).map(bool_op) {
+            Some(b) => Scalar::bool(b, Nullability::Nullable),
```
Comment (robert3005) [inline vortex-array/src/array/constant/compute.rs:106]: This should match the nullability of lhs

File: vortex-array/src/array/constant/compute.rs [hunk 19]
```diff
@@ -59,11 +70,59 @@ impl SearchSortedFn for ConstantArray {
...
+    if constant_array.len() != other.len() {
+        vortex_bail!("Boolean operations aren't supported on arrays of different lengths")
+    }
+
+    if !constant_array.dtype().is_boolean() || !other.dtype().is_boolean() {
+        vortex_bail!("Boolean operations are only supported on boolean arrays")
+    }
+
+    // If the right side is constant
+    if let Some(true) = other.statistics().compute_is_constant() {
```
Comment (robert3005) [inline vortex-array/src/array/constant/compute.rs]: For now we should get the stat only if it's present. Calculation is a bit naive and might be as expensive as running the function.

**Review by robert3005 — CHANGES_REQUESTED** [review]
left a couple of small comments

File: vortex-array/src/array/bool/compute/boolean.rs [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+
+        let array = boolean::or(lhs, rhs)?;
+
+        Ok(ArrayData::from_arrow(&array, true).into_array())
+    }
+}
+
+impl AndFn for BoolArray {
+    fn and(&self, array: &crate::Array) -> vortex_error::VortexResult<crate::Array> {
```
Comment (robert3005) [inline vortex-array/src/array/bool/compute/boolean.rs]: Add use statements for these

File: vortex-array/src/array/varbin/builder.rs (hunk 12 above)
Comment (gatesn) [reply vortex-array/src/array/varbin/builder.rs]: But PrimitiveArray wraps a vortex-buffer, which itself wraps Bytes. So this copy is purely because the right API isn't exposed / isn't used

File: vortex-array/src/array/constant/compute.rs (hunk 18 above)
Comment (AdamGS) [reply vortex-array/src/array/constant/compute.rs:106]: discussed this point - having consistent nullability makes sense here.

File: vortex-array/src/array/varbin/builder.rs (hunk 12 above)
Comment (AdamGS) [reply vortex-array/src/array/varbin/builder.rs]: true, fixing. Added a `from_bytes` function.

File: vortex-array/src/array/bool/compute/boolean.rs (hunk 17 above)
Comment (AdamGS) [reply vortex-array/src/array/bool/compute/boolean.rs:16]: definitely possible, but I think we might be missing some larger abstraction here to solve the whole left/right issue

**Review by robert3005 — APPROVED** [review]
